### ISSUE TITLE
sql: Inconsistent results using savepoint retries

### ISSUE THREAD

**Issue by bdarnell** [cockroachdb/cockroach#6836]
This may be a problem with the cockroach-go repo, but I'm filing it here for more visibility (it also looks more like a server-side bug than a client-side one). The test in `cockroach-go/crdb/tx_test.go` fails for me locally about one in ten times (so does my new port of that code to python):

```
--- FAIL: TestExecuteTx (0.17s)
	tx_test.go:120: expected balances to be restored without error; got acct1=200, acct2=0: %!s(<nil>)
```

In this test, two accounts start with 100 units each, and two racing transactions try to move 100 units from one account to another. If an even number of transactions succeed (as expected), the two accounts end up with a balance of 100. For the test to fail in this way, either one of the transactions had no effect (but still completed without error), or one of the transactions succeeded and was retried anyway (undoing its earlier effect). I've looked at this with `tshark` and I believe it's the former.

**Comment by bdarnell** [comment]
In the python version, I hit a case when both transactions succeeded on the first attempt: https://circleci.com/gh/bdarnell/cockroachdb-python/19

**Comment by tschottdorf** [comment]
Want help investigating this? (Speculative connection to #6747).

**Comment by bdarnell** [comment]
Yeah, I don't have the bandwidth to investigate this myself right now. I'll also speculate that it could be related to the failing kv tests in #6746, #6865 and #6855.

**Comment by tschottdorf** [comment]
Ok, taking a look.

**Comment by tschottdorf** [comment]
Changed the test to run against an explicit local single node for easier debugging. Available at `tschottdorf/acct-transfer`, run via `go test -c . && stress -maxfails 1 -stderr ./crdb.test` (from the `crdb` directory). Will look more tomorrow.

**Comment by tschottdorf** [comment]
Did some SQL-level printf debugging. In the log below,

* 3b35c78b runs both its updates
* 007cfb51 runs both its updates, pushes 3b35c78b, but then also restarts
* 3b35c78b restarts
* 007cfb51 writes (200, 0) and commits
* new player b269a53f shows up and removes 100 from the second account -> (0, -100). Specifically, there's not even a second update corresponding to the other account, but the value seems to commit.

That last item is what's broken. I need to look more into what's going on; for now it looks like there's definitely fishiness in the SQL layer related to 3b35c78b not restarting properly.

I think it's time to use tracing for this. Should be easy enough to look at these in the net/trace endpoint; I didn't think of it before.

```
W160524 22:41:09.073132 sql/update.go:370  UPDATE 3b35c78b [{acct ""} {balance + 100 ""}]: [1 200]
I160524 22:41:09.075854 storage/replica_command.go:1470  range 3316: new leader lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC 406706h41m10.325591522s following replica {0 0 0} 1970-01-01 00:00:00 +0000 UTC 0 [physicalTime=2016-05-25 02:41:09.075827582 +0000 UTC]
I160524 22:41:09.076331 storage/replica_command.go:2011  initiating a split of range=3316 [/Table/2999-/Max) at key /Table/3000
W160524 22:41:09.076470 sql/update.go:370  UPDATE 3b35c78b [{acct ""} {balance - 100 ""}]: [2 0]
W160524 22:41:09.077339 sql/executor.go:448  ERR <nil>
W160524 22:41:09.077554 sql/update.go:370  UPDATE 007cfb51 [{acct ""} {balance + 100 ""}]: [1 200]
W160524 22:41:09.078301 sql/executor.go:448  ERR retry txn "sql/executor.go:420 sql txn" id=3b35c78b key=/Table/3001/1/1/2/1 rw=true pri=0.02554674 iso=SERIALIZABLE stat=PENDING epo=1 ts=1464144069.075584342,1 orig=1464144069.075584342,1 max=1464144069.320935693,0 wto=false rop=false
W160524 22:41:09.078483 sql/executor.go:448  ERR <nil>
W160524 22:41:09.079026 sql/executor.go:448  ERR <nil>
W160524 22:41:09.080497 sql/update.go:370  UPDATE 007cfb51 [{acct ""} {balance - 100 ""}]: [2 0]
W160524 22:41:09.080727 sql/executor.go:448  ERR <nil>
W160524 22:41:09.081677 sql/executor.go:448  ERR retry txn "sql/executor.go:420 sql txn" id=007cfb51 key=/Table/3001/1/1/2/1 rw=true pri=0.02554678 iso=SERIALIZABLE stat=PENDING epo=1 ts=1464144069.075584342,2 orig=1464144069.075584342,2 max=1464144069.322107329,0 wto=false rop=false
W160524 22:41:09.081799 sql/executor.go:448  ERR <nil>
W160524 22:41:09.084491 sql/executor.go:448  ERR <nil>
W160524 22:41:09.084981 sql/update.go:370  UPDATE 007cfb51 [{acct ""} {balance + 100 ""}]: [1 200]
W160524 22:41:09.085816 sql/update.go:370  UPDATE 007cfb51 [{acct ""} {balance - 100 ""}]: [2 0]
W160524 22:41:09.086082 sql/executor.go:448  ERR <nil>
W160524 22:41:09.086467 sql/executor.go:448  ERR <nil>
W160524 22:41:09.086582 sql/executor.go:448  ERR <nil>
I160524 22:41:09.127543 storage/replica_command.go:1470  range 3317: new leader lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC 406706h41m10.377345399s following replica {0 0 0} 1970-01-01 00:00:00 +0000 UTC 0 [physicalTime=2016-05-25 02:41:09.127514663 +0000 UTC]
W160524 22:41:09.127717 sql/update.go:370  UPDATE b269a53f [{acct ""} {balance - 100 ""}]: [2 -100]
```

**Comment by tschottdorf** [comment]
I think what we're seeing is the `SelectClause` which is involved in an `updateNode` not returning a value:

```
08:48:58.233617	 .   478	... UPDATE db21667115703693693.t SET balance = balance + 100 WHERE acct = 1
08:48:58.234627	 .  1009	... UPDATE 1
08:48:58.234629	 .     3	... UPDATE db21667115703693693.t SET balance = balance - 100 WHERE acct = 2
08:48:58.235127	 .   497	... UPDATE 0
```

These updates both happen after a bunch of restarts, and get committed. But clearly all of these updates must affect exactly one row.

**Comment by tschottdorf** [comment]
Got the fix. The update wasn't handling an error returned from the internal
select. The select would return an error (i.e. transaction aborted) and an
empty result set, but the update node would simply treat that as a select
which returned nothing (and no error). Consequently some of the UPDATE
statements turned into no-ops (since they didn't see any rows to update),
resulting in the issues we were seeing (note that this includes negative
balances etc, since the update which wasn't carried out would have failed
in that case). I'm going to upload a PR with a fix and then hand it off to
someone on the SQL team for polishing, regression testing and potentially
an audit of similar gloomy corners of the codebase. We should obviously
have that test in the parallel SQL tests as well (in a form that repros the
issue before the fix).

On Tue, May 24, 2016 at 4:10 PM Ben Darnell <notifications@github.com>
wrote:

> Yeah, I don't have the bandwidth to investigate this myself right now.
> I'll also speculate that it could be related to the failing kv tests in
> #6746 <https://github.com/cockroachdb/cockroach/issues/6746>, #6865
> <https://github.com/cockroachdb/cockroach/issues/6865> and #6855
> <https://github.com/cockroachdb/cockroach/issues/6855>.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/issues/6836#issuecomment-221378995>
>
-- 

-- Tobias

**Comment by tschottdorf** [comment]
here's the diff:

```go
diff --git a/sql/update.go b/sql/update.go
index 1574539..201ebe0 100644
--- a/sql/update.go
+++ b/sql/update.go
@@ -310,6 +310,10 @@ func (u *updateNode) Next() bool {
        }

        if !u.run.rows.Next() {
+               if err := u.run.rows.Err(); err != nil {
+                       u.run.err = err
+                       return false
+               }
                // We're done. Finish the batch.
                err := u.tw.finalize()
                u.run.err = err
```

@vivekmenezes tells me that https://github.com/cockroachdb/cockroach/issues/6747 is exactly the same, but with deleteNode.

**Comment by petermattis** [comment]
This sounds like exactly the same problem @vivekmenezes just uncovered.

**Comment by bdarnell** [comment]
Fixed by #6899 (FYI github only recognizes one issue number per line in the commit message, so this didn't get auto-closed).
